### PR TITLE
IBX-8947: Add 2 cart REST routes; update Cart examples

### PR DESCRIPTION
**Pull request by adriendupuis**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [IBX-8947](https://issues.ibexa.co/browse/IBX-8947)
| Versions      | 4.6, master
| Edition       | All

Add two routes:

- `/api/ibexa/v2/cart/{identifier}/validate` added in ibexa/cart#38 (and asked in [IBX-8947](https://issues.ibexa.co/browse/IBX-8947))
    - Preview: https://ez-systems-developer-documentation--2501.com.readthedocs.build/en/2501/api/rest_api/rest_api_reference/rest_api_reference.html#managing-commerce-carts-cart-validation
- `/api/ibexa/v2/cart/authorize` added in ibexa/cart#80
    - Preview: https://ez-systems-developer-documentation--2501.com.readthedocs.build/en/2501/api/rest_api/rest_api_reference/rest_api_reference.html#managing-commerce-carts-cart-authorization

Add XML examples and update JSON examples.

## Examples generation

Done with 
- Commerce 4.6.11
- DDEV
- ibexa/cart#97
- [xq](https://github.com/kislyuk/xq)

```bash
set -x;

host='http://ddev-ibexa-tmp.ddev.site:8080';
baseUri="$host/api/ibexa/v2";
mimeTypePrefix='application/vnd.ibexa.api';
restRefDir='docs/api/rest_api/rest_api_reference';

for format in 'json' 'xml'; do

parser='jq';
if [ 'xml' = "$format" ]; then
  parser='xq';
fi;

curl "$baseUri/cart/authorize" \
  -c anonymous_cookies.txt \
  -X 'POST' \
  -H "Accept: $mimeTypePrefix.Token+$format" \
> $restRefDir/input/examples/cart/authorize/Token.$format.example;
anonymous_token=`cat $restRefDir/input/examples/cart/authorize/Token.$format.example | $parser -r '.Token.value'`;
if [ 'json' = "$format" ]; then
  echo '' >> $restRefDir/input/examples/cart/authorize/Token.$format.example;
fi;

curl "$baseUri/cart" \
  -b anonymous_cookies.txt -H "X-CSRF-Token: $anonymous_token" \
  -H "Content-Type: $mimeTypePrefix.CartCreate+$format" \
  --data "@$restRefDir/input/examples/cart/POST/CartCreate.$format.example" \
  -H "Accept: application/$format" \
> $restRefDir/input/examples/cart/POST/Cart.$format.example;
identifier=`cat $restRefDir/input/examples/cart/POST/Cart.$format.example | $parser -r '.Cart.identifier'`;
if [ 'json' = "$format" ]; then
  echo '' >> $restRefDir/input/examples/cart/POST/Cart.$format.example;
fi;

curl "$baseUri/cart/$identifier" \
  -b anonymous_cookies.txt -H "X-CSRF-Token: $anonymous_token" \
  -H "Accept: $mimeTypePrefix.Cart+$format" \
> $restRefDir/input/examples/cart/identifier/Cart.$format.example;
if [ 'json' = "$format" ]; then
  echo '' >> $restRefDir/input/examples/cart/identifier/Cart.$format.example;
fi;
if [ '' = "$(diff $restRefDir/input/examples/cart/POST/Cart.$format.example $restRefDir/input/examples/cart/identifier/Cart.$format.example)" ]; then
  rm $restRefDir/input/examples/cart/identifier/Cart.$format.example;
fi;

curl "$baseUri/cart/view" \
  -b anonymous_cookies.txt -H "X-CSRF-Token: $anonymous_token" \
  --data "@$restRefDir/input/examples/cart/view/CartViewInput.$format.example" \
  -H "Content-Type: $mimeTypePrefix.CartViewInput+$format" \
  -H "Accept: $mimeTypePrefix.CartView+$format" \
> $restRefDir/input/examples/cart/view/CartView.$format.example;
if [ 'json' = "$format" ]; then
  echo '' >> $restRefDir/input/examples/cart/view/CartView.$format.example;
fi;

curl "$baseUri/cart/$identifier/validate" \
  -b anonymous_cookies.txt -H "X-CSRF-Token: $anonymous_token" \
  -X 'POST' \
  -H "Accept: $mimeTypePrefix.CartConstraintViolationList+$format" \
> $restRefDir/input/examples/cart/identifier/validate/CartConstraintViolationList.$format.example;
if [ 'json' = "$format" ]; then
  echo '' >> $restRefDir/input/examples/cart/identifier/validate/CartConstraintViolationList.$format.example;
fi;

curl "$baseUri/cart/$identifier" \
  -b anonymous_cookies.txt -H "X-CSRF-Token: $anonymous_token" \
  -X 'DELETE' \
;

rm anonymous_cookies.txt;

done;
exit 0;
```

#### Checklist

- [ ] Text renders correctly
- [ ] Text has been checked with vale
- [ ] Description metadata is up to date
- [ ] Redirects cover removed/moved pages
- [ ] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [ ] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

File: docs/api/rest_api/rest_api_reference/input/examples/cart/POST/CartCreate.xml.example
Comment (adriendupuis) [inline docs/api/rest_api/rest_api_reference/input/examples/cart/POST/CartCreate.xml.example:1]: I'm not able to find a way to add a context [like JSON does](https://github.com/ibexa/documentation-developer/blob/update-cart-rest/docs/api/rest_api/rest_api_reference/input/examples/cart/POST/CartCreate.json.example#L5-L7).

```xml
<CartCreate>
    <name>Default</name>
    <currencyCode>EUR</currencyCode>
    <context>
        <foo>bar</foo>
    </context>
</CartCreate>
```
```xml
?xml version="1.0" encoding="UTF-8"?>
<ErrorMessage media-type="application/vnd.ibexa.api.ErrorMessage+xml">
 <errorCode>500</errorCode>
 <errorMessage>Internal Server Error</errorMessage>
 <errorDescription>Array to string conversion</errorDescription>
 <trace>…</trace>
 <file>/var/www/html/vendor/ibexa/rest/src/lib/Output/Generator/Xml.php</file>
 <line>169</line>
</ErrorMessage>
```

File: docs/api/rest_api/rest_api_reference/input/examples/cart/identifier/validate/CartConstraintViolationList.json.example [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "CartConstraintViolationList": {
+        "_media-type": "application\/vnd.ibexa.api.CartConstraintViolationList+json",
+        "violations": []
```
Comment (adriendupuis) [inline docs/api/rest_api/rest_api_reference/input/examples/cart/identifier/validate/CartConstraintViolationList.json.example:4]: I didn't manage to insert invalid data on purpose.

File: docs/api/rest_api/rest_api_reference/input/ibexa-types.raml [hunk 2]
```diff
@@ -3614,6 +3614,35 @@ CartMetadataUpdate:
...
+        CartConstraintViolationList: CartConstraintViolationList
+
+CartConstraintViolationList:
+    description: 'This class represents a list of cart constraint violations.'
+    type: object
+    properties:
+        violations:
+            type: array
+            items: CartConstraintViolation
+            #TODO: CartConstraintViolationWrapper?
```
Comment (adriendupuis) [inline docs/api/rest_api/rest_api_reference/input/ibexa-types.raml:3630]:
```suggestion
            items: CartConstraintViolation
```

File: docs/api/rest_api/rest_api_reference/input/ibexa-types.raml [hunk 3]
```diff
@@ -3614,6 +3614,34 @@ CartMetadataUpdate:
...
+    properties:
+        CartConstraintViolationList: CartConstraintViolationList
+
+CartConstraintViolationList:
+    description: 'This class represents a list of cart constraint violations.'
+    type: object
+    properties:
+        violations:
+            type: array
+            items: CartConstraintViolation
```
Comment (adriendupuis) [inline docs/api/rest_api/rest_api_reference/input/ibexa-types.raml:3629]: Maybe a CartConstraintViolationWrapper is needed but I didn't manage to product constraint violation to study it.

**Review by konradoboza — APPROVED** [review]

File: docs/api/rest_api/rest_api_reference/input/ibexa-cart.raml [hunk 4]
```diff
@@ -165,3 +170,40 @@ post:
...
+                        application/vnd.ibexa.api.CartConstraintViolationList+xml:
+                            type: CartConstraintViolationList
+                            example: !include examples/cart/identifier/validate/CartConstraintViolationList.xml.example
+                        application/vnd.ibexa.api.CartConstraintViolationList+json:
+                            type: CartConstraintViolationListWrapper
+                            example: !include examples/cart/identifier/validate/CartConstraintViolationList.json.example
+/authorize:
+    post:
+        displayName: Cart Authorization
+        description: Returns the CSRF token required by anonymous user to interact with own cart.
```
Comment (mnocon) [inline docs/api/rest_api/rest_api_reference/input/ibexa-cart.raml:195]:
```suggestion
        description: Returns the CSRF token required by anonymous user to interact with their own cart.
```

**Review by mnocon — APPROVED** [review]
Thanks! The script to generate the responses was definitely useful - especially once I got the right `xq` executable 😄  (https://github.com/mikefarah/yq worked)

File: docs/api/rest_api/rest_api_reference/input/examples/cart/POST/Cart.json.example [hunk 5]
```diff
@@ -1,22 +1,25 @@
 {
     "Cart": {
-        "_media-type": "application/vnd.ibexa.api.Cart+json",
-        "id": 12,
-        "identifier": "fafccc4b-cce1-40d7-8d23-29ad41638581",
+        "_media-type": "application\/vnd.ibexa.api.Cart+json",
```
[the remark this answers is not in the document]
Comment (adriendupuis) [reply docs/api/rest_api/rest_api_reference/input/examples/cart/POST/Cart.json.example:3]: @mnocon 

I checked on https://www.json.org/json-en.html and further on https://ecma-international.org/publications-and-standards/standards/ecma-404/. The standard's PDF states:

> All code points may
be placed within the quotation marks except for the code points that must be escaped: quotation mark
(U+0022), reverse solidus (U+005C), and the control characters U+0000 to U+001F.
> […]
> `\/` represents the solidus character (U+002F).

So escaping the slash (a.k.a solidus) is optional; it doesn't matter.

```
% echo '{ "slash": "\/" }' | jq                                                                                     
{
  "slash": "/"
}
```
`jq` understands it but considers it's not worth keeping it.

We could status for consistency and open a new PR about this. I would vote for removing them all.

**Pull request merged**